r-package/grf/R/forest_summary.R

```diff
@@ -115,69 +122,95 @@ test_calibration <- function(forest) {
+                                   subset = NULL,
+                                   num.trees.for.variance = 500,
+                                   seed = runif(1, 0, .Machine$integer.max)) {
+
   cluster.se <- length(forest$clusters) > 0
-  
+
   if (!("causal_forest" %in% class(forest))) {
     stop("`best_linear_projection` is only implemented for `causal_forest`")
   }
-  
+
   if (is.null(subset)) {
     subset <- 1:length(forest$Y.hat)
   }
-  
+
   if (class(subset) == "logical" & length(subset) == length(forest$Y.hat)) {
     subset <- which(subset)
   }
-  
+
   if (!all(subset %in% 1:length(forest$Y.hat))) {
     stop(paste(
       "If specified, subset must be a vector contained in 1:n,",
       "or a boolean vector of length n."
     ))
   }
-  
+
   clusters <- if (cluster.se) {
     forest$clusters
   } else {
     1:length(forest$Y.orig)
   }
   observation.weight <- observation_weights(forest)
-  
+
   # Only use data selected via subsetting.
+  subset.X.orig <- forest$X.orig[subset, , drop = FALSE]
   subset.W.orig <- forest$W.orig[subset]
   subset.W.hat <- forest$W.hat[subset]
   subset.Y.orig <- forest$Y.orig[subset]
   subset.Y.hat <- forest$Y.hat[subset]
   tau.hat.pointwise <- predict(forest)$predictions[subset]
   subset.clusters <- clusters[subset]
-  subset.weights <- observation.weight[subset]
-  
-  if (min(subset.W.hat) <= 0.01 && max(subset.W.hat) >= 0.99) {
-    rng <- range(subset.W.hat)
-    warning(paste0(
-      "Estimated treatment propensities take values between ",
-      round(rng[1], 3), " and ", round(rng[2], 3),
-      " and in particular get very close to 0 and 1."
-    ))
+  subset.weights.raw <- observation.weight[subset]
+  subset.weights <- subset.weights.raw / mean(subset.weights.raw)
+
+  binary.W <- all(subset.W.orig %in% c(0, 1))
+
+  if (binary.W) {
+    if (min(subset.W.hat) <= 0.01 && max(subset.W.hat) >= 0.99) {
+      rng <- range(subset.W.hat)
+      warning(paste0(
+        "Estimated treatment propensities take values between ",
+        round(rng[1], 3), " and ", round(rng[2], 3),
+        " and in particular get very close to 0 and 1."
+      ))
+    }
+    IPW <- (subset.W.orig - subset.W.hat) / (subset.W.hat * (1 - subset.W.hat))
+    weights <- IPW
+  } else {
+    # Estimate the variance of W given X
+    # Details: https://github.com/grf-labs/grf/issues/608#issuecomment-582747602
+    variance.forest <- regression_forest(subset.X.orig,
+                                        (subset.W.orig - subset.W.hat)^2,
+                                        clusters = subset.clusters,
+                                        num.trees = num.trees.for.variance,
+                                        ci.group.size = 1,
+                                        seed = seed)
+    V.hat <- predict(variance.forest)$predictions
+    debiasing.weights <- subset.weights * (subset.W.orig - subset.W.hat) / V.hat
+    weights <- debiasing.weights
+    # TO ADD: calibrate weights?
```

Comment:
Can you clarify what the idea is for calibrating the weights? Either way we shouldn't leave this comment in

Reply:
The question referred to this code: https://github.com/grf-labs/grf/blob/master/r-package/grf/R/average_partial_effect.R#L101